### ISSUE TITLE
Privesc with fantasy commands.

### ISSUE THREAD

**Issue by 0x00-L0k1** [bbriggs/bitbot#39]
Bitbot by nature will echo out any url title, while this is ok for most situations, in a specific set of circumstances it can be used to take over an IRC channel by an unauthenticated user.

You can reproduce this by registering a channel on an irc server, enabling fantasy commands (!voice, !op..etc) registering bitbot with nickserv, inviting a services bot into the channel, giving bitbot elevated access with chanserv and pasting a url with the title set to "!op badguysname". 

Bitbot with echo the title of the webpage, in this case "!op badguysname" chanserv will recognize the command "!op badguysname" because bitbot is on the access this. 

If you some reason you feel like giving bitbot oper privs...well..you get the point..

**Comment by CriticalCow** [comment]
o/ l0k1

**Comment by bbriggs** [comment]
The injection part of this can be handled in two ways:
1. Any response that contains arbitrary echoes back to chat should be sent as a `NOTICE` instead of a `PRIVMSG`
2. Prefix the `PRIVMSG` response with the requesting user's nick. ie, `foo: Title of Some Site`

**Comment by bbriggs** [comment]
I personally think that `NOTICE`s look ugly in chat, but they do tend to be safer as bots tend to accept commands from `PRIVMSG`

**Comment by NtRaiseHardError** [comment]
> Bitbot by nature will echo out any url title, while this is ok for most situations, in a specific set of circumstances it can be used to take over an IRC channel by an unauthenticated user.
> 
> You can reproduce this by registering a channel on an irc server, enabling fantasy commands (!voice, !op..etc) registering bitbot with nickserv, inviting a services bot into the channel, giving bitbot elevated access with chanserv and pasting a url with the title set to "!op badguysname".
> 
> Bitbot with echo the title of the webpage, in this case "!op badguysname" chanserv will recognize the command "!op badguysname" because bitbot is on the access this.
> 
> If you some reason you feel like giving bitbot oper privs...well..you get the point..

If this is true then the `choose` command can also be used to force this?

`choose !op <target> or !op <target>`

You don't even need to go through the effort of making a webpage.

**Comment by bbriggs** [comment]
Yes, the `choose` command is also vulnerable to this echo problem.
